### PR TITLE
[RFC] First draft of deleted file check

### PR DESCRIPTION
**Pull request by wilsonge**
Pull Request for *some of* Issue #16458 .

### Summary of Changes
Reworks our deleted file script onto something more rudimentary however that takes into account our npm and composer deps are no longer stored in the repo.

#### Things left to do
- [ ] Make the folder paths dynamic at minimum - better to use git to build them ourselves
- [x] Don't trash any installation removed files
- [x] Don't trash search folders or files
- [x] Reverse order the folders list so we delete the nested directories first

### Testing Instructions
This script has been used to generate the existing deleted files list - which are already usable for update - but has had a few modifications since.

Basically in the build create a tmp folder grab a copy of two joomla versions - the 3.x version goes into the a `joomla390` folder and the 4.x into `joomla400` folder (i picked 3.9.13 and 4.0.0 alpha 12) to generate the list in this PR). Then run the script - it should match what's in the com_admin script

### Documentation Changes Required
n/a

### CODE REVIEW

File: build/deleted_file_check.php [hunk 1]
```diff
@@ -1,145 +1,56 @@
...
-
-	exit(1);
-}
-
-/*
- * Constants
- */
-
-const PHP_TAB = "\t";
+// TODO: Make these directories dynamic or clone from git
```
Comment (mbabker) [inline build/deleted_file_check.php]: You definitely want the `--from` and `--to` params from my version of this script here, as you're going to want this script to be able to work without someone needing to arbitrarily set up multiple git clones on their own to run it.  Then, copy the logic from the build script to run the `git archive` command to build your two working directories and do any of the setup steps needed (i.e. NPM and Composer plus cleanup).

The `--from` param, you'll probably want to restrict to real version numbers (tags) instead of any valid commit reference like it does now, because you're going to need it to behave differently based on what version you're doing the comparison from (3.x doesn't have build steps, 4.x does).  I think it's safe to say this version of the script will only be run when doing a comparison to a commit in the 4.x tree, so that restriction isn't needed for the `--to` param.

File: build/deleted_file_check.php [hunk 2]
```diff
@@ -1,145 +1,56 @@
...
 
-asort($deletedFiles);
+asort($filesDifference);
+asort($foldersDifference);
 
 // Write the deleted files list to a file for later reference
-file_put_contents(__DIR__ . '/deleted_files.txt', implode("\n", $deletedFiles));
-
-echo PHP_EOL;
-echo 'There are ' . count($deletedFiles) . ' deleted files in comparison to "' . $options['from'] . '"' . PHP_EOL;
```
Comment (mbabker) [inline build/deleted_file_check.php:145]: Personally, I'd leave an output at the end of the script.  It just helps a little bit in knowing the script's done.

File: build/deleted_file_check.php (hunk 1 above)
Comment (wilsonge) [reply build/deleted_file_check.php]: The problem with git archive is that it gives me all my dev deps as well which I then somehow need to unstage after I've run composer install and npm install. I think we actually need to somehow split the build script out a bit because I need some of the common parts like the files/directories to ignore there

File: build/deleted_file_check.php [hunk 3]
```diff
@@ -90,56 +42,62 @@ function usage($command)
...
 	echo PHP_EOL;
-	echo 'Missing starting commit reference' . PHP_EOL;
-
+	echo 'Missing starting directory' . PHP_EOL;
 	usage($argv[0]);
-
 	exit(1);
 }
 
 // Missing the to reference?  No problem, grab the current HEAD
```
Comment (brianteeman) [inline build/deleted_file_check.php:52]: this comment makes no sense now

Comment (wilsonge) [reply build/deleted_file_check.php:52]: It will go back to being optional eventually. I need to move to the git model michael proposed. This is just adding some placeholder code until i figure out the details

**Comment by wilsonge** [comment]
Pushed a new tweak up - a horrible version of doing a iterator filter (there must be a better way I can deal with the excluded folders other than the prepending which requires separate filters) - but at least it works. This means com_search and installation directories are excluded and there's a simple mechanism for this in the future.

I've also reversed the order of the directories so we delete subdirectories before parent directories

**Comment by wilsonge** [comment]
Just used this to update the latest script.php in 4.0-dev so will be able to use that as a metric of how stable this script is beyond the requirement to be able to download the two branches (for now i just downloaded and unzipped the latest nightlies of 3.10 and 4.0)

**Comment by SharkyKZ** [comment]
How to use this script? I tried `php deleted_file_check.php --from joomla390 --to joomla400` get `Missing ending directory` due to extra colon here https://github.com/joomla/joomla-cms/blob/9d216a27286d82236cfebc6dc9359a7a30ba4512/build/deleted_file_check.php#L39.

**Comment by wilsonge** [comment]
`php build/deleted_file_check.php --from=/Users/george/Downloads/Joomla_3.10.0-dev-Development-Full_Package --to=/Users/george/Downloads/Joomla_4.0.0-beta1-dev-Development-Full_Package`

That's what I did - where the directories are the two nightlies. 

I haven't changed that line from the working J3 version :/ so is this script working for you there?

**Comment by SharkyKZ** [comment]
That works, thanks. I haven't used this script before.

Should templates be deleted? They can contain custom overrides and assets.

**Comment by wilsonge** [comment]
Currently the intention was yes. Because we’re not shipping bootstrap 2 in j4 so they will be broken by default. However I’m willing to listen on it. If people don’t like that decision it’s easy to add it to the exclusion list

**Comment by mbabker** [comment]
> Currently the intention was yes. Because we’re not shipping bootstrap 2 in j4 so they will be broken by default. However I’m willing to listen on it. If people don’t like that decision it’s easy to add it to the exclusion list

Lookup the 2.5 to 3.0 scripts.  IIRC the admin template was fully uninstalled but the frontend template wasn't.  The same should be done for 3.x to 4.0.

**Comment by wilsonge** [comment]
The 2.5 template didn’t break I don’t think because it had no framework under it? Not 100% on that though

**Comment by mbabker** [comment]
Protostar has its own CSS, it’s not using Bootstrap from the media directory.  So that part isn’t busted.  The JavaScript integrations probably are though, not much that can be done for that without a compatibility plugin overloading JHtmlBootstrap to restore BS2 (the same as my old BS3 demo plugin).  Likewise layouts are busted because of the insistence of the default layouts being designed against the default frontend template and not being agnostic.

Still it’s probably be a bad idea to flat out wipe out someone’s template.

**Comment by brianteeman** [comment]
2.5 site and admin templates survived after the update

**Comment by richard67** [comment]
@wilsonge For me this is at least a release blocker due to following reason:

We want to have updateability from beta-n to beta-(n-1).

So latest with beta-2 we need it to be included into our build procedure so the script.php is up to date for each build.

If we want to go one step beyond and support updateability between nightly builds beginning with beta-1, then it even could be a beta-blocker.

Which of these 2 labels should I set? Feedback very welcome.

**Comment by wilsonge** [comment]
I've added an explicit release blocker - but the root issue it's partly solving is already a release blocker.

I don't think it's a beta blocker as we can already use this script - it's just about making it pretty to use

**Comment by richard67** [comment]
@wilsonge I think it should run automatically on every nightly build at the end.

**Comment by wilsonge** [comment]
It can not run automatically - its building array's to paste into script.php. unless we made these contents a standalone file - but that's out of scope of this PR. I'm manually putting the results into script.php right now (as was the case in 3.x too)

**Comment by richard67** [comment]
@wilsonge I thought about making an RFC for using a standalone text file, or 2, 1 for files and 1 for folders, before I posted here. Yes is out of scope of this PR. But if you see a chance for such solution we find maybe time for talk on Glip what could be done how. I am not so familiar with the build process.

**Comment by wilsonge** [comment]
Happy to consider standalone files (obviously would still need a die statement at the top to stop public access but 🤷‍♂ )

**Comment by richard67** [comment]
Now after PR #31804 has been merged, it needs to make sure that files to be **_renamed_** are **_not_** added to the list of files to be **_deleted_**.

It might also make sense to generate a list of files to be renamed, so we have something like a `renamed_files.txt`.

Finally I think it always needs to run the script here in 2 steps:
1. Compare a 4.0 Beta 1 with latest 3.10 Alpha or Beta or whatever release (or maybe even the latest 3.10 nightly).
2. Compare the last 4.0 nightly or a just built one with the last 4.0 release for which the lists have been updated before.

Step 1 might e.g. result in new update SQL scripts for 3.10 which have to be removed when updating to 4.

It needs 2 steps because when comparing latest 4.0 with latest 3.10 only, you miss all those files having been added for 4.0 e.g. with Beta 1 and then later been removed with a later Beta.

**Comment by richard67** [comment]
Hmm, it seems my above suggestion is also not the finally correct one. E.g. file `/administrator/components/com_admin/postinstall/htaccesssvg.php` has been added to 3.10-dev but what not present yet in 4.0 Beta 1, so we would add that file to the list of files to be deleted when comparing those 2 releases.

So it needs to think further 😄

**Comment by dgrammatiko** [comment]
> but what not present yet in 4.0 Beta 1

There's no travel machine yet AFAIK 😄. What I mean is that you can't patch anything to a previous release, so the diff needs to be between 3.10 and the head of the J4 branch. So, the J4 needs a script that will clean up all the files (esp the static assets) from beta1 till the current head and also diff the latest 3.10 release. The previous betas are just broken regarding the removal of some files (esp static assets). In sort let's not complicate things just for shake of writing code. My 2c

**Comment by richard67** [comment]
> In sort let's not complicate things just for shake of writing code. My 2c

That's not my intention. I just want to point out things I find so they are not forgotten.

**Comment by dgrammatiko** [comment]
> I just want to point out things I find so they are not forgotten.

Sure, I think that the only viable solution can only diff the tip of the head the 3.10 with the tip of the head of j4. What happens between beta1 till beta7 (or whatever was the latest release) should be handled as a one-off task (run it once, fix things, delete the script!).

**Comment by richard67** [comment]
Especially in case of 4.0 we have to build an installation package or to run `composer install` with the `no-dev` option and `npm ci`, it's not enough to check just the sources.

**Comment by dgrammatiko** [comment]
> Especially in case of 4.0 we have to build an installation package or to run composer install with the no-dev option and npm ci, it's not enough to check just the sources.

Thankfully Github actions can do that for us automatically, and merge back a file (eg /build/files-diff.json). Then the existing script.php has to only parse this file and add the entries to delete file/folder or rename. Actually it would be a good idea to have this data as a `json` file instead of arrays in a script file. It's easier to automate the additions, removals of entries programmatically


Eg this file: https://github.com/dgrammatiko/sloth-pkg/blob/main/scripts/init.js already has the logic to clone locally (ok in the github actions doker image) any branch. Once it's cloned run composer/npm on j4 and diff the folders. Store the array of files to a file. Have an action to merge back the diff.json. Then patch the script to pick up that data...

**Comment by richard67** [comment]
> Actually it would be a good idea to have this data as a `json` file instead of arrays in a script file. It's easier to automate the additions, removals of entries programmatically

100% agree. It's a pain to maintain the script.php file with these gazillions of lines for these 2 hardcoded monster arrays in it.

**Comment by richard67** [comment]
If we want to detect files which have changed only in case (upper or lowercase mix) of their basename so that they are not added to the list of files to be deleted but to a list of files to be renamed, and if we don't want to limit successful execution of this to a particular OS, e.g. Linux, then we might have to do similar tricks like here to make it safe for any kind of OS:

https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_admin/script.php#L7176-L7218

The reason is that on Windows you can find out what is the actual name of the file, but on iOS you get it in the casing like you ask for it, except when using scandir in PHP in that case, and with a check if the 2 different names have the same inode (like on iOS) or not (i.e. really different files on *nix).

**Comment by dgrammatiko** [comment]
> except when using scandir in PHP in that case

My solution will be Node based, so this won't be a problem. Also since the idea is to automate this with Github actions the OS will be Linux so safe for these operations.

**Comment by richard67** [comment]
Sounds all very good to me. I am not good in JS, but in Git and a few other things, so if I can help with something just let me know.

In the mean time we need to continue with maintaining script.php either with the WiP here or "manually", i.e. we should not do nothing with it while waiting for the good solution. I can help with that, too.

**Comment by dgrammatiko** [comment]
Actually you can help by moving the array of files to a JSON file. A structure that should be ok:
```json
{
  "forRemoval": {
    "files": {
      "3.9.10": [
        "administrator/something/file.php"
      ]
    },
    "folders": {
      "3.2.0": [
        "images/stock"
      ]
    }
  },
  "forRename": {
    "files": {
      "3.9.10": {
        "from": [
          "administrator/something/otherFile.php"
        ],
        "to": [
          "administrator/something/otherfile.php"
        ]
      }
    },
    "folders": {
      "3.2.0": {
        "from": [
          "administrator/Component"
        ],
        "to": [
          "administrator/component"
        ]
      }
    }
  }
}
```

**Comment by dgrammatiko** [comment]
Edit, maybe keep each release to it's own file, similar to what we do for the SQL files. Eg a file named `3.9.10.json` with some content like:
```json
{
  "forRemoval": {
    "files":  [
        "administrator/something/file.php"
      ],
    "folders": [
        "images/stock"
      ]
  },
  "forRename": {
    "files": {
      "from": [
        "administrator/something/otherFile.php"
      ],
      "to": [
        "administrator/something/otherfile.php"
      ]
    },
    "folders": {
      "from": [
        "administrator/Component"
      ],
      "to": [
        "administrator/component"
      ]
    }
  }
}
```

Then in the script.php scan the dir (let's say named `redundant`) read each file and json parse it. Loop the json and create arrays for flies and folders to be removed, renamed. Then map the arrays to the particular function (delete, rename). Joba done. This is also maintainable...

**Comment by dgrammatiko** [comment]
FWIW I just setup a rough CI https://github.com/dgrammatiko/jdiff with gihub actions and get the redundant files: https://github.com/dgrammatiko/jdiff/blob/main/Redundant.json 

Forking this all you have to do is run the action and copy-paste the redundant.json (I mean if we go with the idea of moving the data to json files)

**Comment by dgrammatiko** [comment]
@wilsonge @richard67 how can I verify that the collection of files https://github.com/dgrammatiko/jdiff/blob/main/Redundant.json and folders https://github.com/dgrammatiko/jdiff/blob/main/RedundantFolders.json is what we expect to be for the J3 to J4 upgrade?

**Comment by richard67** [comment]
> @wilsonge @richard67 how can I verify that the collection of files https://github.com/dgrammatiko/jdiff/blob/main/Redundant.json and folders https://github.com/dgrammatiko/jdiff/blob/main/RedundantFolders.json is what we expect to be for the J3 to J4 upgrade?

By comparing 4.0 Beta 1 with the latest nightly of 3.10-dev.

**Comment by dgrammatiko** [comment]
Sure, but are there files that although it shouldn't survive the upgrade we want to keep around (eg protostar, beez templates)

**Comment by richard67** [comment]
> Sure, but are there files that although it shouldn't survive the upgrade we want to keep around (eg protostar, beez templates)

I think the templates are deleted ... about which other files to keep I don't know. We might really have to wait for George for an answer to that.

**Comment by dgrammatiko** [comment]
Ok, anyways on the https://github.com/dgrammatiko/jdiff whenever you create and merge a PR will create the diffs for files and folders (based on what **EXISTS** on J4 and J3.10 latest branches.

**Comment by richard67** [comment]
I'll have a look on it but it may take some time.
